src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1517 @@
+                          ark_mem->tempv2, ark_mem->user_data);
+    step_mem->nfe++;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    /* Estimate the local error and compute its weighted RMS norm */
+    cvals[0] = p8;
+    Xvecs[0] = ark_mem->yn;
+    cvals[1] = -p8;
+    Xvecs[1] = ark_mem->ycur;
+    cvals[2] = p4 * ark_mem->h;
+    Xvecs[2] = ark_mem->fn;
+    cvals[3] = p4 * ark_mem->h;
+    Xvecs[3] = ark_mem->tempv2;
+
+    retval = N_VLinearCombination(4, cvals, Xvecs, ark_mem->tempv1);
+    if (retval != 0) { return (ARK_VECTOROP_ERR); }
+
+    *dsmPtr = N_VWrmsNorm(ark_mem->tempv1, ark_mem->ewt);
+    lsrkStep_DomEigUpdateLogic(ark_mem, step_mem, *dsmPtr);
+  }
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepRKL:
+
+  This routine serves the primary purpose of the LSRKStepRKL module:
+  it performs a single RKL step (with embedding, if possible).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKL(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype* cvals;
+  sunrealtype hmax, w1, bjm1, bjm2, mus, bj, ajm1, cjm1, temj, cj, mu, nu;
+  const sunrealtype p8 = SUN_RCONST(0.8), p4 = SUN_RCONST(0.4);
+  N_Vector* Xvecs;
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  cvals = step_mem->cvals;
+  Xvecs = step_mem->Xvecs;
+
+  /* Compute dominant eigenvalue and update stats */
+  if ((step_mem->newdomeig))
+  {
+    retval = lsrkStep_ComputeNewDomEig(ark_mem, step_mem);
+    if (retval != ARK_SUCCESS) { return (retval); }
+    step_mem->ndomeigupdates++;
+  }
+
+  /* determine the number of required stages */
+  int ss;
+  for (ss = 1; ss <= step_mem->stagemaxlimit; ss++)
+  {
+    if ((SUNSQR(ss) + ss - 2) >= 2 * (SUNRabs(ark_mem->h) * step_mem->sprad))
+    {
+      step_mem->reqstages = SUNMAX(ss, 2);
+      break;
+    }
+    if (ss == step_mem->stagemaxlimit)
+    {
+      hmax = SUN_RCONST(0.95) * (SUNSQR(ss) + ss - 2) / (2.0 * step_mem->sprad);
```

Comment:
To that end, we need to create a new double in the step memory and assign its value from `hadapt_mem->safety` when both of their memories are available (probably in `lsrkStep_SetDefaults`). Do you think we should have a new `step_mem->safety` kind of parameter, or keep it in this way?

Reply:
On second though, `sprad` already has the `domeigsfty` applied, so I wonder if the `0.95` should be outright removed.

Reply:
I believe they serve to different goals. `domeigsfty`  is for user provided domeig while `0.95` is there to avoid hitting the `step_mem->stagemaxlimit` again and to have slightly less `step_mem->reqstages` than the limittig value. We can make this quantity equal to `domeigsfty` or assign a new name and an adjustable value. That won't be totally irrelevant. However, I think we shouldn't totally remove it.